cachi2/core/package_managers/rpm/main.py

```diff
@@ -0,0 +1,194 @@
+
+        with rpmfile.open(file_path) as rpm:
+            vendor = rpm.headers.get("vendor", b"").decode()
+            vendor = quote(vendor.lower())
+
+            name = rpm.headers.get("name", b"").decode()
+            version = rpm.headers.get("version", b"").decode()
+            release = rpm.headers.get("release", b"").decode()
+            arch = rpm.headers.get("arch", b"").decode()
+            download_url = quote(file_metadata["url"])
+
+            # https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm
+            # https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairsa
+            purl = (
+                f"pkg:rpm{'/' if vendor else ''}{vendor}/{name}@{version}-{release}"
+                f"?arch={arch}{'&epoch=' if epoch else ''}{epoch}&download_url={download_url}"
+            )
+
+            if file_metadata["checksum"] is None:
+                missing_hash_in_file = file_path.name
+                properties = [
+                    Property(name="cachi2:missing_hash:in_file", value=missing_hash_in_file)
+                ]
+            else:
+                properties = []
+
+            components.append(
+                Component(
+                    name=name,
+                    version=version,
+                    purl=purl,
+                    properties=properties,
+                )
+            )
+
+
+def process_packages(path: Path) -> None:
+    """
+    Search structure for all repoid dirs and create repository metadata
+    out of its RPMs (and SRPMs).
+    """
+    package_dir = path.joinpath(DEFAULT_PACKAGE_DIR)
+    for arch in package_dir.iterdir():
+        if not arch.is_dir():
+            continue
+        for repoid in arch.iterdir():
+            if not repoid.is_dir():
+                continue
+            createrepo(repoid.name, repoid.absolute())
+
+
+def createrepo(repoid: str, repodir: Path) -> None:
+    """Execute the createrepo utility."""
+    log.info(f"Creating repository metadata for repoid '{repoid}': {repodir}")
+    cmd = ["createrepo_c", str(repodir)]
+    log.debug("$ " + shlex.join(cmd))
+    stdout = run_cmd(cmd, params={})
+    log.debug(stdout)
+
+
+def _generate_repofiles(path: RootedPath, lockfile: RedhatRpmsLock) -> list[ProjectFile]:
+    """
+    Search structure and generate repofile for each arch in its dir.
+    Repofile contains all arch's repoids. Skip sources dir that contains only SRPMs.
+    """
+    project_files: list[ProjectFile] = []
+    package_dir = path.path.joinpath(DEFAULT_PACKAGE_DIR)
+    for arch in package_dir.iterdir():
+        if not arch.is_dir():
+            continue
+        log.debug(f"Preparing repofile for arch '{arch.name}'")
+        template = ""
+        for repoid in arch.iterdir():
+            if not repoid.is_dir():
+                continue
+            localpath = Path("$output_dir", DEFAULT_PACKAGE_DIR, arch.name, repoid.name)
```

Comment:
What's $output_dir?

Reply:
This relates to the comment:
https://github.com/containerbuildsystem/cachi2/pull/502#discussion_r1537251809
During `inject-files` it is evaluated to `--for-output-dir` argument value.

Reply:
More precisely, it's related to the `ProjectFile` template wrapping class.